File: dotNet/Unit-1/FileValidatorActor.cs
﻿using System.IO;
using Akka.Actor;

namespace WinTail
{
    class FileValidatorActor : UntypedActor
    {
        private readonly IActorRef _consoleWriterActor;

        public FileValidatorActor(IActorRef consoleWriterActor)
        {
            _consoleWriterActor = consoleWriterActor;
        }

        protected override void OnReceive(object message)
        {
            var msg = message as string;
            if (string.IsNullOrEmpty(msg))
            {
                _consoleWriterActor.Tell(new Message.NullInputError("Input was blank. Please try again\n"));
                Sender.Tell(new Message.ContinueProcessing());
            }
            else
            {
                if(isFileURI(msg))
                {
                    _consoleWriterActor.Tell(new Message.InputSuccess(string.Format("Starting processing {0}", msg)));
                    Context.ActorSelection("akka://myActorSystem/user/tailCoordinatorActor").Tell(new TailCoordinatiorActor.StartTail(msg, _consoleWriterActor));
                }
                else
                {
                    _consoleWriterActor.Tell(new Message.ValidationError(string.Format("{0} is not an existing URI on disk", msg)));
                    Sender.Tell(new Message.ContinueProcessing());
                }
            }
        }

        private bool isFileURI(string path)
        {
            return File.Exists(path);
        }
    }
}

File: scala/unit-1/src/main/scala/com/example/actors/FileValidatorActor.scala
package com.example.actors

import akka.actor.{ Actor, ActorLogging }
import akka.actor.ActorRef
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.FileSystems
import akka.actor.Props

object FileValidator {
  def props(consoleWriter: ActorRef):Props = Props(new FileValidatorActor(consoleWriter))
}


class FileValidatorActor(consoleWriter: ActorRef) extends Actor with ActorLogging {

  def receive = {
    case line: String =>
      if (noMessage(line)) nullInputError
      else if (isValid(line)) valid(line)
      else validationError(line)
      sender ! ConsoleMessages.ContinueProcessing
  }

  private def validationError(file:String) = consoleWriter ! ConsoleMessages.ValidationError(s"$file isn't a valid file URI")
  private def nullInputError = consoleWriter ! ConsoleMessages.NullInputError("No input received.")
  private def valid(file:String): Unit = {
    consoleWriter ! ConsoleMessages.InputSuccess(s"Starting to process $file")
    context.actorSelection("akka://Main/user/app/tailCoordinator") ! TailCoordinator.Messages.StartTail(file, consoleWriter)
  }
  private def noMessage(line: String) = line.isEmpty()
  private def end(line: String) = line.toLowerCase() == ConsoleMessages.endCommand
  private def isValid(filePath: String): Boolean = Files.exists(FileSystems.getDefault().getPath(filePath))
}
